File: code/jvm/src/main/kotlin/pt/isel/leic/daw/explodingbattleships/domain/Square.kt
package pt.isel.leic.daw.explodingbattleships.domain

import java.lang.IllegalArgumentException

/**
 * Represents a square
 * @property row the square's row
 * @property column the square's column
 */
data class Square(
    val row: Char,
    val column: Int
) {
    /**
     * Get the square placed above the current square
     */
    fun up() =
        Square(row - 1, column)

    /**
     * Get the square placed below the current square
     */
    fun down() =
        Square(row + 1, column)

    /**
     * Get the square placed to the left of the current square
     */
    fun left() =
        Square(row, column - 1)

    /**
     * Get the square placed to the right of the current square
     */
    fun right() =
        Square(row, column + 1)

    /**
     * Get a string with the row and column of the square
     */
    override fun toString() = "$row$column"
}

/**
 * Converts the string to square or null
 */
fun String.toSquareOrNull() =
    try { Square(first(), subSequence(1, lastIndex + 1).toString().toInt()) } catch (e: Exception) { null }

/**
 * Converts the string to a square or throws an exception
 */
fun String.toSquare() = toSquareOrNull() ?: throw IllegalArgumentException("Invalid square")

File: code/js/src/domain/Square.ts
export type Square = {
    row: string,
    column: number
}

export function up(square: Square): Square { 
    return {row: String.fromCharCode(square.row.charCodeAt(0) - 1), column: square.column}
}
export function down(square: Square): Square {
     return {row: String.fromCharCode(square.row.charCodeAt(0) + 1), column: square.column}
}
export function left(square: Square): Square {
    return {row: square.row, column: square.column - 1}
}
export function right(square: Square): Square {
    return {row: square.row, column: square.column + 1}
}

export function surroundingSquares(square: Square) {
    return [
        left(up(square)),
        up(square),
        right(up(square)),
        left(square),
        square,
        right(square),
        left(down(square)),
        down(square),
        right(down(square))
    ]
}

export function squareToString(square: Square) {
    return `${square.row}${square.column}`
}